AppCenter/AppCenter/Internals/Util/MSEncrypter.m

```diff
@@ -117,52 +65,150 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
+  NSDate *expiration = [[NSDate date] dateByAddingTimeInterval:kMSEncryptionKeyLifetimeInSeconds];
+  NSString *expirationIso = [MSUtility dateToISO8601:expiration];
+
+  // Format is {keyTag}/{expiration as iso}.
+  NSString *keyMetadata = [@[ newKeyTag, expirationIso ] componentsJoinedByString:kMSEncryptionMetadataInternalSeparator];
+  [[MSUserDefaults shared] setObject:keyMetadata forKey:kMSEncryptionKeyMetadataKey];
+}
+
+- (NSData *)getKeyWithKeyTag:(NSString *)keyTag {
+  NSData *key;
+  @synchronized(self) {
+
+    // Read inside of a synchronized block because NSDictionary is not thread safe.
+    key = [self.keys objectForKey:keyTag];
+  }
+
+  // If key is not cached, try loading it from Keychain.
+  if (!key) {
+    NSString *stringKey = [MSKeychainUtil stringForKey:keyTag];
+
+    // If key is not saved in Keychain, create one and save it.
+    if (!stringKey) {
+      @synchronized(classLock) {
+
+        // Recheck if the key has been written from another thread.
+        stringKey = [MSKeychainUtil stringForKey:keyTag];
+        if (!stringKey) {
+          key = [[self class] generateAndSaveKeyWithTag:keyTag];
+        }
+      }
+    } else {
+      key = [[NSData alloc] initWithBase64EncodedString:stringKey options:0];
+    }
+    @synchronized(self) {
+
+      // Write inside of a synchronized block because NSDictionary is not thread safe.
+      [self.keys setObject:key forKey:keyTag];
+    }
+  }
+  return key;
 }
 
-+ (NSData *)loadKeyFromKeychainWithTag:(NSString *)keyTag {
-  NSData *keyData = nil;
-  NSString *stringKey = [MSKeychainUtil stringForKey:keyTag];
-  if (stringKey) {
-    keyData = [[NSData alloc] initWithBase64EncodedString:stringKey options:0];
++ (NSData *_Nullable)performCryptoOperation:(CCOperation)operation
+                                      input:(NSData *)input
+                       initializationVector:(NSData *)initializationVector
+                                        key:(NSData *)key {
+  NSData *result;
+
+  // Create a buffer whose size is at least one block plus 1. This is not needed for decryption, but it works.
+  size_t outputBufferSize = [input length] + kCCBlockSizeAES128 + 1;
+  uint8_t *outputBuffer = malloc(outputBufferSize * sizeof(uint8_t));
+  size_t numBytesNeeded = 0;
+  CCCryptorStatus status =
+      CCCrypt(operation, kMSEncryptionAlgorithm, kCCOptionPKCS7Padding, [key bytes], kMSEncryptionKeySize, [initializationVector bytes],
+              [input bytes], input.length, outputBuffer, outputBufferSize, &numBytesNeeded);
+  if (status != kCCSuccess) {
+
+    // Do not print the status; it is a security requirement that specific crypto errors are not printed.
+    MSLogError([MSAppCenter logTag], @"Error performing encryption or decryption.");
+  } else {
+    result = [NSData dataWithBytes:outputBuffer length:numBytesNeeded];
+    if (!result) {
+      MSLogWarning([MSAppCenter logTag], @"Could not create NSData object from encrypted or decrypted bytes.");
```

Comment:
Why is this a warning not an error?

Reply:
Essentially the same message was a warning before. Could change to error.